src/main/java/com/accelerex/tasks_manager/serviceImpl/UserServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+    private final PasswordEncoder passwordEncoder;
+    private final JWTUtil jwtUtil;
+    private final EmailService emailService;
+
+    @Override
+    public List<UserDto> getAllUsers() {
+        log.info("Fetching all users...");
+        List<User> userList = userRepository.findAll();
+        List<UserDto> responseList = new ArrayList<>();
+        for (User user : userList) {
+            UserDto response = new UserDto();
+            response.setFirstName(user.getFirstName());
+            response.setLastName(user.getLastName());
+            response.setEmail(user.getEmail());
+//            response.setRole(user.getRole());
+            responseList.add(response);
+        }
+        log.info("Fetched {} users", responseList.size());
+        return responseList;
+    }
+
+    @Override
+    public UserDto getUserById(Long userId) {
+        log.info("Fetching user by ID: {}", userId);
+        Optional<User> optionalUser = userRepository.findById(userId);
+        if (optionalUser.isPresent()) {
+            User user = optionalUser.get();
+            UserDto userDto = new UserDto();
+            userDto.setId(user.getId());
+            userDto.setFirstName(user.getFirstName());
+            userDto.setLastName(user.getLastName());
+            userDto.setEmail(user.getEmail());
+//            userDto.setRole(user.getRole());
+            return userDto;
+        } else {
+            throw new UserNotFoundException("User with ID " + userId + " not found");
+        }
+    }
+
+
+    @Override
+    public UserDto signUpUser(UserSignUpDto signUpDto) {
+        User newUser = new User();
+        newUser.setPassword(passwordEncoder.encode(signUpDto.getPassword()));
+        newUser.setSecurityQuestion(signUpDto.getSecurityQuestion());
+        newUser.setSecurityAnswer(signUpDto.getSecurityAnswer());
+        User savedUser = userRepository.save(newUser);
+
+        UserDto userDto = new UserDto();
+        userDto.setId(savedUser.getId());
+        userDto.setFirstName(savedUser.getFirstName());
+        userDto.setLastName(savedUser.getLastName());
+        userDto.setEmail(savedUser.getEmail());
+
+        return userDto;
+    }
+
+    @Override
+    public UserDto adminSignUp(UserDto dto) {
+        Optional<User> optionalUser = userRepository.findUserByEmail(dto.getEmail());
+        if (optionalUser.isEmpty()) {
+            User user = new User();
+            user.setEmail(dto.getEmail());
+            user.setFirstName(dto.getFirstName());
+            user.setLastName(dto.getLastName());
+            user.setPassword(passwordEncoder.encode(dto.getPassword()));
+            user.setDisabled(true);
+            Optional<UserRole> role = userRoleRepository.findById(dto.getRoleId());
+            if (dto.getRoleId() != null && role.isPresent()) {
+                user.setRole(role.get());
+            }
+            userRepository.save(user);
+        } else {
+            throw new RuntimeException("Email already exist");
```

Comment:
do you want to create a custom exception say `EmailAlreadyExistException` or a generic one `ApplicationException` then handle it in you controllerAdvice.